hail/python/hail/methods/statgen.py

```diff
@@ -1697,6 +1697,244 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
+        n_groups = math.ceil(a.shape[0] / group_size)
+        groups = []
+        for i in range(n_groups):
+            start = i * group_size
+            if i == n_groups - 1:
+                end = a.shape[0]
+            else:
+                end = (i + 1) * group_size
+            groups.append(a[start:end, :])
+        return groups
+
+    def concatBlocked(A):
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        ht = ndarray_to_table([big_mat])
+        return ht
+
+    def concatToNumpy(A):
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        return big_mat
+
+    def ndarray_to_table(chunked_arr):
+        structs = [hl.struct(row_group_number = idx, ndarray = block)
+                   for idx, block in enumerate(chunked_arr)]
+        ht = hl.Table.parallelize(structs)
+        ht = ht.key_by('row_group_number')
+        return ht
+
+    def block_product(left, right):
+        product = left @ right     
+        n_rows, n_cols = product.shape
+        return hl.struct(
+            shape=product.shape,
+            block=hl.range(hl.int(n_rows * n_cols)).map(
+                lambda absolute: product[absolute % n_rows, absolute // n_rows]))
+
+    # takes in output of block_product
+    def block_aggregate(prod):
+        shape = prod.shape
+        block = prod.block
+        return hl.nd.from_column_major(
+            hl.agg.array_sum(block),
+            hl.agg.take(shape, 1)[0])
+
+    # returns flat array
+    def to_column_major(ndarray):
+        n_rows, n_cols = ndarray.shape
+        return hl.range(hl.int(n_rows * n_cols)).map(
+            lambda absolute: ndarray[absolute % n_rows, absolute // n_rows])
+
+    def matmul_rowblocked_nonblocked(A, B):
+        temp = A.annotate_globals(mat = B)
+        temp = temp.annotate(ndarray = temp.ndarray @ temp.mat)
+        temp = temp.select(temp.ndarray)
+        temp = temp.drop(temp.mat)
+        return temp
+
+    # def matmul_colblocked_rowblocked(A, B):
+    #     temp = A.transmute(ndarray = block_product(A.ndarray.transpose(), B[A.row_group_number].ndarray))
+    #     result_arr_sum = temp.aggregate(block_aggregate(temp.ndarray))
+    #     return result_arr_sum
+
+    def matmul_colblocked_rowblocked(A, B):
+        temp = A.transmute(ndarray = A.ndarray.transpose() @ B[A.row_group_number].ndarray)
+        return temp.aggregate(hl.agg.ndarray_sum(temp.ndarray))
+
+    def computeNextH(A, H):
+        nextG = matmul_colblocked_rowblocked(A, H)
+        return matmul_rowblocked_nonblocked(A, nextG)
+
+
+    # Algorithm
+
+    def hailBlanczos(A, G, k, l, q, block_size):
+    
+        # assert l > k
+        # assert n <= m
+        
+        Hi = matmul_rowblocked_nonblocked(A, G)
+        npH = concatToNumpy(Hi)
+        for j in range(0, q):
+            Hj = computeNextH(A, Hi)
+            npH = np.concatenate((npH, concatToNumpy(Hj)), axis=1)
+            Hi = Hj
```

Comment:
We can improve this loop. Right now, it does the following: 

Before the loop:
1. You multiply A by G once.
2. You collect that locally.

In the loop:
3. To compute Hj, you multiply AT by Hi, collect that down into a small matrix `nextG`, and then multiply A by `nextG`. 
4. Then you collect Hj locally.

We want something closer to Patrick's notes:

1. Start from `A`. It has one field, `ndarray`. 
2. Do:
```
temp = A.annotate(H_i = A.ndarray @ G)
temp = temp.annotate(G_i_intermediate = temp.ndarray.T @ temp.H_i)
result = temp.aggregate(hl.struct(Hi_chunks = hl.agg.collect(temp.H_i), G_i = hl.agg.ndarray_sum(temp.G_i))
localized_H_i = np.concat(result.Hi_chunks, axis = 1)
G_i = result.G_i
```
3. Append your `localized_H_i` to the `H` you're building up.
4. Repeat this process using the new `G_i`.

Reply:
Note that A never changes, it's reused the same way every time.